bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/api/dto/clip2/helper/Setters.java

```diff
@@ -341,7 +349,60 @@ public static Resource setResource(Resource target, Resource source) {
                 targetTimedEffects.setDuration(duration);
             }
         }
-
         return target;
     }
+
+    /**
+     * Merge on/dimming/color fields from light and grouped light resources.
+     * Subsequent resources will be merged into the first one.
+     * Full state resources are not supported by this method.
+     */
+    public static Collection<Resource> mergeLightResources(Collection<Resource> resources) {
+        Map<String, Resource> resourceIndex = new HashMap<>();
+        Iterator<Resource> iterator = resources.iterator();
+        while (iterator.hasNext()) {
+            Resource resource = iterator.next();
+            String id = resource.getId();
+
+            if (resource.hasFullState()) {
+                throw new IllegalStateException("Resource " + id + " has full state, this is not expected");
+            }
+
+            Resource indexedResource = resourceIndex.get(id);
+            if (indexedResource == null) {
+                indexedResource = resource;
+                resourceIndex.put(id, indexedResource);
+            }
+
+            if (!LIGHT_TYPES.contains(resource.getType()) || !resource.hasHSBField()
+                    || resource.equals(indexedResource)) {
+                continue;
+            }
+
+            OnState onState = resource.getOnState();
+            if (onState != null) {
+                indexedResource.setOnState(onState);
+                resource.setOnState(null);
+            }
+            Dimming dimming = resource.getDimming();
+            if (dimming != null) {
+                indexedResource.setDimming(dimming);
+                resource.setDimming(null);
+            }
+            ColorXy colorXy = resource.getColorXy();
+            if (colorXy != null) {
+                indexedResource.setColorXy(colorXy);
+                resource.setColorXy(null);
+            }
+
+            if (!resource.hasAnyRelevantField()) {
+                iterator.remove();
```

Comment:
@jlaur I think you are trying to do the “garbage collection” twice — once by iterator.remove inside the loop, and again via removeIf outside the loop.

.. and if you decide to do it via removeIf, then it may be that an iterator is not required, and the loop could be executed more simply via a forEach syntax.

Reply:
Correct. I went with the additional `iterator.remove()` for performance reasons, because in situations where it's possible to decide to remove the object straight away, this will avoid an additional traversal of the collection.

The performance gain might be negligible, so if you feel strongly for removing this optimization, I'm fine with that.

Reply:
I have no strong feelings about the optimisation. But if the remove is done within the loop then the external removeIf is dead code so it can be deleted.

Reply:
I suggest to delete the removeIf line. And as you say we can modify hasAnyRelevantField in future as required.

.. maybe hasAnyRelevantField should be called hasAnyLightRelevantField ??

Reply:
I have removed the removal of irrelevant resources.

> .. maybe hasAnyRelevantField should be called hasAnyLightRelevantField ??

I think it's fine as is. Currently it supports only **light** and **grouped_light**, but it can easily be extended to support other types when needed.

Reply:
The current method name is not really self documenting (relevant to what) .. a fully self documenting method name would be hasFieldsNotRelevantToMergingofPureLightStateInLightOrGroupedLightResources 

.. perhaps you can explain it in the Java doc?